### ISSUE TITLE
Error evaluating DeepRenewalProcess model

### ISSUE THREAD

**Issue by dmitra79** [awslabs/gluon-ts#2171]
## Description

I trained DeepRenewalProcessModel successfully, and then tried running evaluation (on a Rolling Dataset), but ran into the error below. I have trained and evaluated ex. DeepAR models with the exact same dataset and evaluation code without an issue.
Would appreciate any suggestions on what the problem could be.

## Error message or code output

```
TypeError                                 Traceback (most recent call last)
<ipython-input-11-b27ba6261c6f> in <module>
      1 all_ds_res=[]
      2 for (n, dataset) in zip(ds_names,eval_datasets):
----> 3     agg_metrics, inst_metrics = compute_metrics(dataset, predictor)
      4     d={}
      5     d['predictor']=experiment.params['id']

<ipython-input-10-0bf47719dd63> in compute_metrics(dataset, predictor)
      9     )
     10     evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9], num_workers=None)
---> 11     return evaluator(ts_it, forecast_it)

~/conda/derv2/lib/python3.7/site-packages/gluonts/evaluation/_base.py in __call__(self, ts_iterator, fcst_iterator, num_series)
    225                 mp_pool.join()
    226             else:
--> 227                 for ts, forecast in it:
    228                     rows.append(self.get_metrics_per_ts(ts, forecast))
    229 

~/conda/derv2/lib/python3.7/site-packages/tqdm/std.py in __iter__(self)
   1178 
   1179         try:
-> 1180             for obj in iterable:
   1181                 yield obj
   1182                 # Update and possibly print the progressbar.

~/conda/derv2/lib/python3.7/site-packages/gluonts/model/renewal/_predictor.py in predict(self, dataset, num_samples, **kwargs)
    121                 freq=self.freq,
    122                 output_transform=self.output_transform,
--> 123                 num_samples=num_samples,
    124             )
    125 

~/conda/derv2/lib/python3.7/site-packages/gluonts/model/forecast_generator.py in __call__(self, inference_data_loader, prediction_net, input_names, freq, output_transform, num_samples, **kwargs)
    170         **kwargs
    171     ) -> Iterator[Forecast]:
--> 172         for batch in inference_data_loader:
    173             inputs = [batch[k] for k in input_names]
    174             outputs = predict_to_numpy(prediction_net, inputs)

~/conda/derv2/lib/python3.7/site-packages/gluonts/transform/_base.py in __iter__(self)
    103     def __iter__(self) -> Iterator[DataEntry]:
    104         yield from self.transformation(
--> 105             self.base_dataset, is_train=self.is_train
    106         )
    107 

~/conda/derv2/lib/python3.7/site-packages/gluonts/transform/_base.py in __call__(self, data_it, is_train)
    126                 yield self.map_transform(data_entry.copy(), is_train)
    127             except Exception as e:
--> 128                 raise e
    129 
    130     @abc.abstractmethod

~/conda/derv2/lib/python3.7/site-packages/gluonts/transform/_base.py in __call__(self, data_it, is_train)
    124         for data_entry in data_it:
    125             try:
--> 126                 yield self.map_transform(data_entry.copy(), is_train)
    127             except Exception as e:
    128                 raise e

~/conda/derv2/lib/python3.7/site-packages/gluonts/transform/_base.py in map_transform(self, data, is_train)
    139 
    140     def map_transform(self, data: DataEntry, is_train: bool) -> DataEntry:
--> 141         return self.transform(data)
    142 
    143     @abc.abstractmethod

~/conda/derv2/lib/python3.7/site-packages/gluonts/transform/_base.py in transform(self, data)
    158 
    159     def transform(self, data: DataEntry) -> DataEntry:
--> 160         return self.func(data.copy())
    161 
    162 

~/conda/derv2/lib/python3.7/site-packages/gluonts/mx/batchify.py in batchify(data, ctx, dtype, variable_length, is_right_pad)
     98             is_right_pad=is_right_pad,
     99         )
--> 100         for key in data[0].keys()
    101     }
    102 

~/conda/derv2/lib/python3.7/site-packages/gluonts/mx/batchify.py in <dictcomp>(.0)
     98             is_right_pad=is_right_pad,
     99         )
--> 100         for key in data[0].keys()
    101     }
    102 

~/conda/derv2/lib/python3.7/site-packages/gluonts/mx/batchify.py in stack(data, ctx, dtype, variable_length, is_right_pad)
     79         data = mx.nd.array(data, dtype=dtype, ctx=ctx)
     80     elif isinstance(data[0], (list, tuple)):
---> 81         return list(stack(t, ctx=ctx) for t in zip(*data))
     82     return data
     83 

~/conda/derv2/lib/python3.7/site-packages/gluonts/mx/batchify.py in <genexpr>(.0)
     79         data = mx.nd.array(data, dtype=dtype, ctx=ctx)
     80     elif isinstance(data[0], (list, tuple)):
---> 81         return list(stack(t, ctx=ctx) for t in zip(*data))
     82     return data
     83 

~/conda/derv2/lib/python3.7/site-packages/gluonts/mx/batchify.py in stack(data, ctx, dtype, variable_length, is_right_pad)
     77         return mx.nd.stack(*data)
     78     if isinstance(data[0], np.ndarray):
---> 79         data = mx.nd.array(data, dtype=dtype, ctx=ctx)
     80     elif isinstance(data[0], (list, tuple)):
     81         return list(stack(t, ctx=ctx) for t in zip(*data))

~/conda/derv2/lib/python3.7/site-packages/mxnet/ndarray/utils.py in array(source_array, ctx, dtype)
    144         return _sparse_array(source_array, ctx=ctx, dtype=dtype)
    145     else:
--> 146         return _array(source_array, ctx=ctx, dtype=dtype)
    147 
    148 

~/conda/derv2/lib/python3.7/site-packages/mxnet/ndarray/ndarray.py in array(source_array, ctx, dtype)
   3359                 source_array = np.array(source_array, dtype=dtype)
   3360             except:
-> 3361                 raise TypeError('source_array must be array like object')
   3362 
   3363     if source_array.shape == ():

TypeError: source_array must be array like object
```

## Environment
-  Operating system: linux
-  Python version: 3.7.0
-  GluonTS version: 0.9.4
-  MXNet version: 1.7.0

**Comment by lostella** [comment]
@dmitra79 it seems from the error trace that the issue is in the shape of some arrays that are being stacked to form a batch. But it’s hard to say more.

Do you have a minimal code example that reproduces the issue? For example with small dummy data using a `ListDataset` is fine.

**Comment by pfolch** [comment]
Hi!
I had the same problem. In order to reproduce it, I have created a dummy example, which I have attached below:

```
import numpy as np
import pandas as pd
from gluonts.dataset.common import ListDataset
from gluonts.model.renewal import DeepRenewalProcessEstimator
from gluonts.mx import Trainer
from gluonts.mx.distribution.neg_binomial import NegativeBinomialOutput
from gluonts.mx.distribution.poisson import PoissonOutput

NUM_TS = 10
MAX_LENGTH_TS = 100
MIN_LENGTH_TS = 5
MAX_VALUE_TS = 3
prediction_length = 4
with_fdr = True

target = []
for i in range(NUM_TS):
    target.append(np.random.randint(0, MAX_VALUE_TS, np.random.randint(MIN_LENGTH_TS, MAX_LENGTH_TS)))

indices = pd.date_range(start="2021-01-01", end="2021-01-31")

train_time_series_dicts = []
test_time_series_dicts = []
for i, time_series in enumerate(target):
    dict_test = {"target": time_series,
                 "start": indices[0],
                 "fsc": [i, i % 2]}
    dict_train = {"target": time_series[:-prediction_length],
                  "start": indices[0],
                  "fsc": [i, i % 2]}
    if with_fdr:
        fdr = [np.random.randint(0, 3, len(time_series))]
        dict_test.update({"fdr": fdr})
        dict_train.update({"fdr": [fdr[0][:-prediction_length]]})

    test_time_series_dicts.append(dict_test)
    train_time_series_dicts.append(dict_train)

test_ds = ListDataset(test_time_series_dicts, freq='W')
train_ds = ListDataset(train_time_series_dicts, freq='W')

trainer = Trainer(epochs=5)
estimator = DeepRenewalProcessEstimator(
    prediction_length=prediction_length,
    context_length=10,
    num_cells=40,
    num_layers=2,
    trainer=trainer,
    freq="W",
    interval_distr_output=NegativeBinomialOutput(),
    size_distr_output=PoissonOutput()
)

predictor = estimator.train(train_ds, test_ds)

preds = list(predictor.predict(test_ds))
```


If the with_fdr flag is set to False, the error does not occur, or at least, in the trials I have made, it has not occurred. On the other hand, with the flag set to True, the error is easily reproduced.
Thanks!

**Comment by dmitra79** [comment]
Thanks @pfolch ! I can confirm the same issue with this example

**Comment by lostella** [comment]
@dmitra79 @pfolch this is fixed by #2184 and we'll be part of the upcoming bugfix releases. In the meantime, you can test the fix using the `dev` branch.
